File: Sources/Services/FileTree/FileTreeBuilder.swift
// MARK: - FileTreeBuilder
// Builds a FileTreeNode hierarchy from the file system.
// Filters excluded directories and annotates git status.
// macOS 14+, Swift 5.10

import Foundation

/// Builds a hierarchical ``FileTreeNode`` tree from a directory on disk.
///
/// Excluded directories (node_modules, .git, .build, DerivedData, etc.)
/// are filtered out. Git status annotations are applied from a provided
/// ``GitStatus`` instance.
enum FileTreeBuilder {

    /// Directories that are always excluded from the tree.
    private static let excludedNames: Set<String> = PathConstants.excludedDirectoryNames

    /// Build a file tree for the given directory.
    ///
    /// - Parameters:
    ///   - root: Root directory URL.
    ///   - gitStatus: Optional git status for annotating file nodes.
    ///   - maxDepth: Maximum recursion depth (nil = unlimited).
    /// - Returns: Array of root-level ``FileTreeNode`` entries.
    static func buildTree(
        at root: URL,
        gitStatus: GitStatus? = nil,
        maxDepth: Int? = nil
    ) -> [FileTreeNode] {
        let fm = FileManager.default

        // Build a lookup map of relative path -> git status.
        var gitMap: [String: GitFileStatus] = [:]
        if let gs = gitStatus {
            for file in gs.stagedFiles { gitMap[file.path] = file.status }
            for file in gs.unstagedFiles { gitMap[file.path] = file.status }
            for file in gs.untrackedFiles { gitMap[file.path] = file.status }
        }

        // R-14: Track visited canonical paths to detect symlink loops.
        var visitedPaths = Set<String>()

        let context = BuildContext(
            root: root,
            fileManager: fm,
            gitMap: gitMap,
            maxDepth: maxDepth
        )

        return buildLevel(
            at: root,
            context: context,
            currentDepth: 0,
            visitedPaths: &visitedPaths
        )
    }

    // MARK: - Private

    /// Invariant inputs shared across every recursive ``buildLevel`` call.
    private struct BuildContext {
        let root: URL
        let fileManager: FileManager
        let gitMap: [String: GitFileStatus]
        let maxDepth: Int?
    }

    private static func gitRelativePath(for url: URL, root: URL) -> String? {
        let rootPaths = [root.path, root.resolvingSymlinksInPath().path]
        let filePaths = [url.path, url.resolvingSymlinksInPath().path]

        for filePath in filePaths {
            for rootPath in rootPaths {
                let normalizedRootPath = rootPath.hasSuffix("/") ? String(rootPath.dropLast()) : rootPath
                let prefix = normalizedRootPath + "/"
                if filePath.hasPrefix(prefix) {
                    return String(filePath.dropFirst(prefix.count))
                }
            }
        }

        return nil
    }

    private static func buildLevel(
        at directory: URL,
        context: BuildContext,
        currentDepth: Int,
        visitedPaths: inout Set<String>
    ) -> [FileTreeNode] {
        if let max = context.maxDepth, currentDepth >= max {
            return []
        }

        // R-14: Resolve symlinks and check for loops to prevent infinite recursion.
        let canonicalPath = directory.resolvingSymlinksInPath().path
        guard !visitedPaths.contains(canonicalPath) else {
            return []
        }
        visitedPaths.insert(canonicalPath)

        // Prefetch only `.isDirectoryKey`: it is the sole resource value read
        // below, and enumeration caches it on each returned URL. `.isHiddenKey`
        // was prefetched previously but never consumed — dropping it avoids an
        // unused per-entry stat during directory enumeration.
        guard let contents = try? context.fileManager.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsPackageDescendants]
        ) else {
            return []
        }

        var directories: [FileTreeNode] = []
        var files: [FileTreeNode] = []

        for url in contents {
            let name = url.lastPathComponent

            // Skip excluded directories and hidden files.
            if excludedNames.contains(name) { continue }

            // Served from the enumeration prefetch cache — no additional stat.
            let resourceValues = try? url.resourceValues(forKeys: [.isDirectoryKey])
            let isDirectory = resourceValues?.isDirectory ?? false

            if isDirectory {
                let children = buildLevel(
                    at: url,
                    context: context,
                    currentDepth: currentDepth + 1,
                    visitedPaths: &visitedPaths
                )
                let entry = DirectoryEntry(
                    path: url,
                    children: children,
                    isExpanded: false
                )
                directories.append(.directory(entry))
            } else {
                let relativePath = gitRelativePath(for: url, root: context.root)
                let gitStatus = relativePath.flatMap { context.gitMap[$0] }
                let entry = FileEntry(path: url, gitStatus: gitStatus)
                files.append(.file(entry))
            }
        }

        // Sort: directories first (alphabetical), then files (alphabetical).
        directories.sort { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
        files.sort { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }

        return directories + files
    }
}

File: shared/src/commonMain/kotlin/studio/vibe/shared/service/filetree/FileTreeBuilder.kt
package studio.vibe.shared.service.filetree

import studio.vibe.shared.constants.PathConstants
import studio.vibe.shared.contract.PersistenceStore
import studio.vibe.shared.model.DirectoryEntry
import studio.vibe.shared.model.FileEntry
import studio.vibe.shared.model.FilePath
import studio.vibe.shared.model.FileTreeNode
import studio.vibe.shared.model.GitFile
import studio.vibe.shared.model.GitFileStatus
import studio.vibe.shared.model.GitStatus

/**
 * Builds a [FileTreeNode] hierarchy from a directory on disk.
 *
 * Excluded directories (node_modules, .git, .build, DerivedData, etc.) are
 * filtered out. Git status annotations are applied from an optional [GitStatus].
 *
 * Symlink loop detection is done by tracking canonical path strings that the
 * platform persistence layer resolves; each path is visited at most once.
 */
class FileTreeBuilder(private val persistence: PersistenceStore) {

    /** Directories that are always excluded from the tree. */
    private val excludedNames: Set<String> = PathConstants.excludedDirectoryNames

    /**
     * Build a file tree for the given directory.
     *
     * @param root      Root directory path.
     * @param gitStatus Optional git status for annotating file nodes.
     * @param maxDepth  Maximum recursion depth (null = unlimited).
     * @return List of root-level [FileTreeNode] entries, directories first then files,
     *   each group sorted alphabetically (case-insensitive).
     */
    suspend fun buildTree(
        root: FilePath,
        gitStatus: GitStatus? = null,
        maxDepth: Int? = null,
    ): List<FileTreeNode> {
        // Build a lookup map of relative path -> git status.
        val gitMap = mutableMapOf<String, GitFileStatus>()
        if (gitStatus != null) {
            fun addFiles(files: List<GitFile>) {
                for (file in files) gitMap[file.path] = file.status
            }
            addFiles(gitStatus.stagedFiles)
            addFiles(gitStatus.unstagedFiles)
            addFiles(gitStatus.untrackedFiles)
        }

        val visitedPaths = mutableSetOf<String>()
        return buildLevel(
            directory = root,
            root = root,
            gitMap = gitMap,
            currentDepth = 0,
            maxDepth = maxDepth,
            visitedPaths = visitedPaths,
        )
    }

    // ---------------------------------------------------------------------------
    // Private
    // ---------------------------------------------------------------------------

    private suspend fun buildLevel(
        directory: FilePath,
        root: FilePath,
        gitMap: Map<String, GitFileStatus>,
        currentDepth: Int,
        maxDepth: Int?,
        visitedPaths: MutableSet<String>,
    ): List<FileTreeNode> {
        if (maxDepth != null && currentDepth >= maxDepth) return emptyList()

        // Symlink loop detection: use the canonical path as the identity key.
        // PersistenceStore exposes the path string directly; the platform
        // implementation is responsible for resolving symlinks before listing.
        val canonicalPath = directory.path
        if (canonicalPath in visitedPaths) return emptyList()
        visitedPaths.add(canonicalPath)

        val contents = try {
            persistence.listDirectory(directory)
        } catch (_: Exception) {
            return emptyList()
        }

        val directories = mutableListOf<FileTreeNode>()
        val files = mutableListOf<FileTreeNode>()

        for (childPath in contents) {
            val name = childPath.name

            // Skip excluded directories and hidden files.
            if (name in excludedNames) continue

            val isDir = try {
                persistence.isDirectory(childPath)
            } catch (_: Exception) {
                false
            }

            if (isDir) {
                val children = buildLevel(
                    directory = childPath,
                    root = root,
                    gitMap = gitMap,
                    currentDepth = currentDepth + 1,
                    maxDepth = maxDepth,
                    visitedPaths = visitedPaths,
                )
                val entry = DirectoryEntry(
                    path = childPath,
                    children = children,
                    isExpanded = false,
                )
                directories.add(FileTreeNode.Directory(entry))
            } else {
                // Compute path relative to root for git status lookup.
                val rootPrefix = root.path.trimEnd('/') + "/"
                val relativePath = if (childPath.path.startsWith(rootPrefix)) {
                    childPath.path.removePrefix(rootPrefix)
                } else {
                    childPath.path
                }
                val status = gitMap[relativePath]
                val entry = FileEntry(path = childPath, gitStatus = status)
                files.add(FileTreeNode.File(entry))
            }
        }

        // Sort: directories first (alphabetical), then files (alphabetical).
        directories.sortWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.name })
        files.sortWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.name })

        return directories + files
    }
}
